Add more assessor tools

jsf9k opened the pull request:
## 🗣 Description ##

This pull request adds more assessor tools to the Kali AMI.  It also improves the installation of a few existing tools.

## 💭 Motivation and context ##

Resolves cisagov/cool-system-internal#66.

## 🧪 Testing ##

I built a new AMI for our staging COOL environment and verified that it contains the new tools.

## ✅ Pre-merge checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.
- [x] Build and test a new staging AMI with these changes.
- [x] Finalize version.

## ✅ Post-merge checklist ##

- [ ] Build a new production AMI.

jsf9k committed:
Add AutoBlue, BananaPhone, and BOFs

jsf9k committed:
Add donut and DotNetToJScript

jsf9k committed:
Add cifs-utils package

jsf9k committed:
Install Empire from GitHub instead of via the Kali package manager

jsf9k committed:
Add EXCELntDonut

jsf9k committed:
Install kerberoast from GitHub rather than the Kali package manager

jsf9k committed:
Explicitly add nfs-common to the list of Kali packages to install

jsf9k committed:
Add Terminator

jsf9k committed:
Add the Kali package testssl.sh

jsf9k committed:
Add PCredz

jsf9k committed:
Add ScareCrow

jsf9k committed:
Install SecLists via GitHub instead of via the Kali package

jsf9k committed:
Install crackmapexec via GitHub instead of via the Kali package manager

jsf9k committed:
Install mimikatz via GitHub instead of via the Kali package manager

jsf9k committed:
Install Responder via GitHub instead of via the Kali package manager

jsf9k committed:
Install sqlmap via GitHub instead of via the Kali package manager

jsf9k committed:
Move PCredz installation tasks into their own file

jsf9k committed:
Remove unnecessary quotes

jsf9k committed:
Install nightly build of Metasploit rather than the Kali package

jsf9k committed:
Alphabetize Cobalt Strike task

jsf9k committed:
Add Visual Studio Code

jsf9k committed:
Add Firefox Dev Edition

jsf9k committed:
Add the requests package as a requirement for checkpwnedemails

jsf9k committed:
Specify that demiguise is a Python 2 tool

jsf9k committed:
Move installation of donut to its own tasks file

Donut requires the C code to be built before the Python module can be
built, so it needs some extra steps.

jsf9k committed:
Specify that ftpenum is a Python 2 tool

jsf9k committed:
Install Hasher from GitHub instead of via pip

jsf9k committed:
Install mitm6 from GitHub instead of via pip

jsf9k committed:
Specify that morphHTA is a Python 2 tool

jsf9k committed:
Add glob2 and configparser as dependencies for SimplyTemplate

jsf9k committed:
Force the creation of a Python 3 venv for sqlmapproject

jsf9k committed:
Drop the BloodHound Ansible role and instead create a separate task file

The installation is a bit more complicated now since the assessors
want the GitHub version instead of the version from the Kali package
manager.  We also add installation of the Python ingestor for
BloodHound.

jsf9k committed:
Install neo4j from the official apt repo and start the service at boot

The version of neo4j available by default in Kali does not include the
SystemD service file, so we need to install the package from a
different apt repo.

jsf9k committed:
Bump version from 0.4.10 to 0.4.11

jsf9k committed:
Bump version from 0.4.11 to 0.4.11-rc.1

jsf9k committed:
Add the vnc user to the kaboxer group

The kaboxer group is created when firefox-developer-edition-en-us-kbx
package is installed, and the vnc user must be a member of .the group
to run Firefox Developer Edition.

jsf9k committed:
Bump version from 0.4.11-rc.1 to 0.4.11-rc.2

dav3r commented on src/assessor_tools.yml:
Fix typo and use consistent capitalization with the rest of this file:
```suggestion
    # firefox-developer-edition-en-us-kbx package, and the VNC user
    # must be a member of this group to run Firefox Developer Edition.
    - name: Add VNC user to kaboxer group
```

dav3r reviewed (approved):
Powerful wizardry! 🧙

jsf9k committed:
Fix a typo and make capitalization consistent

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj commented on src/assessor_tools.yml:
```diff
@@ -65,10 +103,31 @@
+          - --strip-components=1
+    - role: assessment_tool
+      vars:
+        archive_src: "https://github.com/FortyNorthSecurity/EXCELntDonut/\
```
Looking at this repo it looks like it has a really janky dependency on https://github.com/Accenture/CLRvoyance where the `install.sh` script clones that repo and moves it into the directory structure. The `EXCELntDonut/CLRvoyance` directory is empty both when I clone the repo and when I download the tarball.

jsf9k replied:
There are likely other janky dependencies hidden in these tools as well.  My general rule has been to install dependencies where the tool repo does things in a standard way and otherwise let the assessors handle the installation.  In this case, though, I created commit 0d8dfc8 to handle the additional dependency.

mcdonnnj replied:
Unfortunately the change does not fully duplicate the logic in the `install.sh` script, but I'm not sure if you want to implement the `mv EXCELntDonut/CLRvoyance/CLRvoyance/* EXCELntDonut/CLRvoyance` at the end (the full command is
```
git clone https://github.com/Accenture/CLRvoyance EXCELntDonut/CLRvoyance && mv EXCELntDonut/CLRvoyance/CLRvoyance/* EXCELntDonut/CLRvoyance
```
). I'm not sure how you want to proceed with this, but since you mentioned your general rule I'm ok with reverting this addition if you don't want to fool with fully replicating the `install.sh` script.

jsf9k replied:
I un-`tar` the [Accenture/CLRvoyance](https://github.com/Accenture/CLRvoyance) tarball into `/tools/EXCELntDonut/EXCELntDonut/CLRvoyance` with `--strip-components=1`, which I think reproduces the effect of the `mv` command.  What am I missing?

mcdonnnj replied:
The `install.sh` script would be moving the contents of `/tools/EXCELntDonut/CLRvoyance/CLRvoyance` up to `/tools/EXCELntDonut/CLRvoyance`. It's a dirty, dirty hack to add `CLRvoyance` as a module within the `EXCELntDonut` package.

jsf9k replied:
Please have a look at commit 952b62d, which I believe does the correct thing.

mcdonnnj replied:
Makes sense. Sorry for the extra work!

jsf9k replied:
No problem.  It was necessary work.

mcdonnnj commented on src/assessor_tools.yml:
```diff
@@ -187,18 +263,19 @@
           - --strip-components=1
     - role: assessment_tool
       vars:
         archive_src: https://github.com/vysecurity/morphHTA/tarball/master
```
I looked at the script in this repo, saw the `import` lines ended in semicolons, and promptly closed that tab.

jsf9k replied:
Can we resolve this conversation?

mcdonnnj commented on src/assessor_tools.yml:
```diff
@@ -242,6 +319,31 @@
+          - dnspython
+          - impacket
+          - ldap3
+          - readline
```
I think these packages are also required for this tool. `pycryptodome` is a drop-in replacement for `pycrypto` and uses the `Crypto` namespace. `pycryptodomex` is a library implementation of the same code that lives under the `Cryptodome` namespace. I believe both of them are in use.

Note: I used [pipreqs](https://github.com/bndr/pipreqs) to gather information about requirements.

```suggestion
          - pyasn1
          - pycryptodome
          - pycryptodomex
          - readline
          - six
```

mcdonnnj commented on src/assessor_tools.yml:
```diff
@@ -242,6 +319,31 @@
+        archive_src: https://github.com/lgandx/Responder/tarball/master
+        install_dir: /tools/Responder
+        pip_packages:
+          - crypto
```
I don't believe this library is needed. The references to `Crypto` are I believe for `PyCrypto` which, based on https://github.com/lgandx/Responder/issues/144#issuecomment-780180422, we should install `PyCryptodome` as an equivalent/replacement.

```suggestion
```

mcdonnnj commented on src/assessor_tools.yml:
```diff
@@ -242,6 +319,31 @@
+          - impacket
+          - ldap3
+          - readline
+        python2: yes
```
Just wanted to note that this tool has a bunch of scripts on top of the main `Responder` script, and a lot of them require Python 3. Are there any that explicitly require Python 2?

jsf9k replied:
You're right.  This tool is not a Python 2 tool.
```suggestion
```

mcdonnnj commented on src/assessor_tools.yml:
While I'm skimming past~

```suggestion
    - name: Install Aquatone
```

mcdonnnj commented on src/bloodhound.yml:
```diff
@@ -0,0 +1,72 @@
+---
+- name: Install and configure neo4j
```
```suggestion
- name: Install and configure Neo4j
```

mcdonnnj commented on src/bloodhound.yml:
```diff
@@ -0,0 +1,72 @@
+---
+- name: Install and configure neo4j
+  block:
+    # The version of neo4j available by default in Kali does not
```
```suggestion
    # The version of Neo4j available by default in Kali does not
```

mcdonnnj commented on src/bloodhound.yml:
```diff
@@ -0,0 +1,72 @@
+    # package from a different repo.
+    - name: Configure apt to use official neo4j apt repo
+      block:
+        - name: Add key for official neo4j apt repo
```
```suggestion
        - name: Add key for official Neo4j apt repo
```

mcdonnnj commented on src/bloodhound.yml:
```diff
@@ -0,0 +1,72 @@
+        - name: Add key for official neo4j apt repo
+          ansible.builtin.apt_key:
+            url: https://debian.neo4j.com/neotechnology.gpg.key
+        - name: Add official apt repo for neo4j
```
```suggestion
        - name: Add official apt repo for Neo4j
```

mcdonnnj commented on src/bloodhound.yml:
```diff
@@ -0,0 +1,72 @@
+          ansible.builtin.apt_repository:
+            repo: deb https://debian.neo4j.com stable 4.3
+
+    - name: Install neo4j
```
```suggestion
    - name: Install Neo4j
```

mcdonnnj commented on src/bloodhound.yml:
```diff
@@ -0,0 +1,72 @@
+        name:
+          - neo4j
+
+    # BloodHound will not work unless you change the neo4j password
```
```suggestion
    # BloodHound will not work unless you change the Neo4j password
```

mcdonnnj commented on src/bloodhound.yml:
```diff
@@ -0,0 +1,72 @@
+
+    # BloodHound will not work unless you change the neo4j password
+    # from the default value.
+    - name: Set neo4j password
```
```suggestion
    - name: Set Neo4j password
```

mcdonnnj commented on src/bloodhound.yml:
```diff
@@ -0,0 +1,72 @@
+      ansible.builtin.systemd:
+        daemon_reload: true
+
+    - name: Start neo4j service at boot
```
```suggestion
    - name: Start Neo4j service at boot
```

mcdonnnj commented on src/bloodhound.yml:
```diff
@@ -0,0 +1,72 @@
+      ansible.builtin.package:
+        name:
+          - npm
+    - name: Install electron-packager
```
```suggestion
    - name: Install Electron Packager
```

mcdonnnj commented on src/donut.yml:
```diff
@@ -0,0 +1,23 @@
+---
+# We have to grab the code, build the C code, and then we can build
+# the python extension
```
```suggestion
# the Python extension
```

mcdonnnj commented on src/donut.yml:
```diff
@@ -0,0 +1,23 @@
+---
+# We have to grab the code, build the C code, and then we can build
+# the python extension
+- name: Install donut
```
```suggestion
- name: Install Donut
```

mcdonnnj commented on src/donut.yml:
```diff
@@ -0,0 +1,23 @@
+    unarchive_extra_opts:
+      - --strip-components=1
+
+- name: Build donut
```
```suggestion
- name: Build Donut
```

mcdonnnj commented on src/donut.yml:
```diff
@@ -0,0 +1,23 @@
+  ansible.builtin.make:
+    chdir: /tools/donut
+
+- name: Create venv and build donut python extension
```
```suggestion
- name: Create venv and build Donut Python extension
```

mcdonnnj commented on src/assessor_tools.yml:
```suggestion
    - name: Install Donut
```

mcdonnnj commented on src/assessor_tools.yml:
```suggestion
    - name: Install BloodHound
```

mcdonnnj reviewed (changes requested):
Thanks for slogging through all of this. Aside from the capitalization suggestions I had some concerns about two of the tools being installed that I'd like to get resolved.

jsf9k committed:
Make capitalization consistent

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Fix Responder dependencies

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Remove Python 2 specification for Responder

Co-authored-by:  Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add installation of EXCELntDonut dependency

EXCELntDonut has a janky dependency on Accenture/CLRvoyance.  The
install.sh script that is included with EXCELntDonut clones
Accenture/CLRvoyance into a subdirectory, so I duplicate that behavior
in the Ansible playbook.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Move EXCELntDonut installation to a separate playbook

The installation of the Accenture/CLRvoyance dependency is a bit more
complicated than I first thought.

mcdonnnj commented on src/assessor_tools.yml:
```diff
@@ -283,9 +377,22 @@
         archive_src: "https://github.com/SimplySecurity/SimplyTemplate/\
         tarball/master"
         install_dir: /tools/SimplyTemplate
+        pip_packages:
```
Do we care about the following system Python packages their `setup.sh` script installs?

* python-tk
* python-pmw
* python-imaging

jsf9k replied:
Please see commit 2df5ba0.

I don't believe it is necessary to install `tk` via `pip`, since `tkinter` (AKA `Tkinter` in Python 2) has been built into Python since ~2008.

mcdonnnj commented on src/bloodhound.yml:
```diff
@@ -0,0 +1,72 @@
+        chdir: /tools/BloodHound
+        cmd: npm run linuxbuild
+
+- name: Install Python ingestor for Bloodhound
```
```suggestion
- name: Install Python ingestor for BloodHound
```

mcdonnnj reviewed (approved):
Thanks for working on my previous feedback. I had one capitalization suggestion I missed last time. I also had a question, but it does not block my approval.

jsf9k committed:
Make capitalization consistent

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add more dependencies for SimplySecurity/SimplyTemplate

These Python dependencies are installed via the Setup.sh script in the
SimplySecurity/SimplyTemplate repository.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Bump version from 0.4.11-rc.2 to 0.4.11
